backend/src/connector/mod.rs

```diff
@@ -1 +1,267 @@
+use self::{
+    api::{APIError, ApiNoContext},
+    client::Client,
+    context::ClientContext,
+};
+use crate::{
+    models::{
+        bob::NodeName,
+        shared::{BobConnectionData, Hostname, RequestTimeout, XSpanIdString},
+    },
+    prelude::*,
+};
+use api::ContextWrapperExt;
+use axum::headers::{authorization::Basic, Authorization};
+use error_stack::ResultExt;
+use frunk::hlist;
+use hyper::StatusCode;
+use std::{collections::HashMap, sync::Arc};
+use thiserror::Error;
 
+pub mod api;
+pub mod client;
+pub mod context;
+pub mod dto;
+
+pub type ApiInterface = dyn ApiNoContext<ClientContext> + Send + Sync;
+
+#[derive(Debug, Error)]
+pub enum ClientError {
+    #[error("couldn't init http client")]
+    InitClient,
+    #[error("couldn't probe connection to the node")]
+    Inaccessible,
+    #[error("permission denied")]
+    PermissionDenied,
+    #[error("no client found found for requested resource")]
+    NoClient,
+    #[error("bad address: no port")]
+    NoPort,
+    #[error("bad address: couldn't parse hostname")]
+    BadAddress,
+    #[error("Can't form hyper request")]
+    CantFormRequest,
+}
+
+/// HTTP Connector constructor
+#[derive(Debug)]
+pub struct Connector;
+
+impl Connector {
+    /// Alows building a HTTP(S) connector. Used for instantiating clients with custom
+    /// connectors.
+    #[must_use]
+    pub const fn builder() -> Builder {
+        Builder {}
+    }
+}
+
+/// Builder for HTTP(S) connectors
+#[derive(Debug)]
+pub struct Builder {}
+
+impl Builder {
+    /// [Stub] Use HTTPS instead of HTTP
+    #[must_use]
+    pub const fn https(self) -> HttpsBuilder {
+        HttpsBuilder {}
+    }
+
+    /// Build a HTTP connector
+    #[must_use]
+    pub fn build(self) -> hyper::client::connect::HttpConnector {
+        hyper::client::connect::HttpConnector::new()
+    }
+}
+
+// TODO
+/// [Stub] Builder for HTTPS connectors
+#[derive(Debug)]
+pub struct HttpsBuilder {}
+
+impl HttpsBuilder {
+    pub fn build(self) {
+        unimplemented!()
+    }
+}
+
+#[derive(Clone)]
+pub struct BobClient {
+    /// Bob's hostname
+    hostname: Hostname,
+
+    // NOTE: Can (and should) the API interface mutate?..
+    /// Connection,
+    main: Arc<ApiInterface>,
```

Comment:
If you see the situation, when some other type can be here (not the `Client`), then generic type would be sufficient

Reply:
The problem with generics is that `BobClient` is used all over the program, and type annotations will be required for all the auth layer for example (or rewrite it fully). That's seems... cumbersome. `dyn` keyword removes type annotation and opens a way for implementing other clients inside of the same `BobClient` instance
But I agree that generics is a better way to do this